File: problems/54-Spiral-Matrix/solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* spiralOrder(int** matrix, int matrixSize, int* matrixColSize, int* returnSize){
    int *result = (int *)malloc(matrixSize * matrixColSize[0] * sizeof(int));
    memset(result, 0, matrixSize * matrixColSize[0] * sizeof(int));
    *returnSize = matrixSize * matrixColSize[0];
    
    int x = 0, y = 0, index = 0, direction = 1;
    
    result[index++] = matrix[x][y];
    matrix[x][y] = 10000;
    while(true) {
        if(direction == 1) {
            if(y < matrixColSize[0] - 1 && matrix[x][y + 1] != 10000) {
                result[index++] = matrix[x][y + 1];
                matrix[x][y + 1] = 10000;
                y++;
            } else if(x < matrixSize - 1 && matrix[x + 1][y] != 10000) {
                result[index++] = matrix[x + 1][y];
                matrix[x + 1][y] = 10000;
                x++;
                direction = 2;
            } else {
                break;
            }
        } else if(direction == 2) {
            if(x < matrixSize - 1 && matrix[x + 1][y] != 10000) {
                result[index++] = matrix[x + 1][y];
                matrix[x + 1][y] = 10000;
                x++;
            } else if(y > 0 && matrix[x][y - 1] != 10000) {
                result[index++] = matrix[x][y - 1];
                matrix[x][y - 1] = 10000;
                direction = 3;
                y--;
            } else {
                break;
            }
        } else if(direction == 3) {
            if(y > 0 && matrix[x][y - 1] != 10000) {
                result[index++] = matrix[x][y - 1];
                matrix[x][y - 1] = 10000;
                y--;
            } else if(x > 0 && matrix[x - 1][y] != 10000) {
                result[index++] = matrix[x - 1][y];
                matrix[x - 1][y] = 10000;
                x--;
                direction = 4;
            } else {
                break;
            }
        } else {
            if(x > 0 && matrix[x - 1][y] != 10000) {
                result[index++] = matrix[x - 1][y];
                matrix[x - 1][y] = 10000;
                x--;
            } else if(y < matrixColSize[0] - 1 && matrix[x][y + 1] != 10000) {
                result[index++] = matrix[x][y + 1];
                matrix[x][y + 1] = 10000;
                y++;
                direction = 1;
            } else {
                break;
            }
        }
    }
    
    return result;
}

File: problems/54-Spiral-Matrix/solution.go
func spiralOrder(matrix [][]int) []int {
	row, r, col, c, direction := len(matrix), 0, len(matrix[0]), 0, 1
	dummy := 1000

	result := make([]int, 0, row*col)
	for {
		result = append(result, matrix[r][c])
		matrix[r][c] = dummy

		if direction == 1 {
			if c < col-1 && matrix[r][c+1] != dummy {
				c++
			} else if r < row-1 && matrix[r+1][c] != dummy {
				r++
				direction = 2
			} else {
				break
			}
		} else if direction == 2 {
			if r < row-1 && matrix[r+1][c] != dummy {
				r++
			} else if c > 0 && matrix[r][c-1] != dummy {
				c--
				direction = 3
			} else {
				break
			}
		} else if direction == 3 {
			if c > 0 && matrix[r][c-1] != dummy {
				c--
			} else if r > 0 && matrix[r-1][c] != dummy {
				r--
				direction = 4
			} else {
				break
			}
		} else {
			if r > 0 && matrix[r-1][c] != dummy {
				r--
			} else if c < col-1 && matrix[r][c+1] != dummy {
				c++
				direction = 1
			} else {
				break
			}
		}
	}

	return result
}

